### DISCUSSION TITLE
Delete the file in dokan

### DISCUSSION THREAD

**Discussion by zqlustc** [dokan-dev/dokany#974] [Q&A] [answered]
Now I have a file in the dokan mount point, for example, named test.txt. Right-click to delete the test1 file it will delete test.txt with no prompt, but it maybe failed in the dokan fuse. After I refesh the parent DIR of test.txt. It appears again. I wrote a hack in DokanDispatchCreate function in dokan sys as following.
```
    disposition = (irpSp->Parameters.Create.Options >> 24) & 0x000000ff;
    if (irpSp->Parameters.Create.Options & FILE_DELETE_ON_CLOSE) {
      status = STATUS_INVALID_PARAMETER;
      __leave;
    }
```

But it can not avoid the problem. Is it true that windows have deleted the file view before call create dokan sys ops? @kyanha

**Comment by Liryna** [comment]
The system ask you if the file can be deleted before actually doing it (making it disappear in explorer).
Have you been using Procmon to find what way is used to remove the file ? `FILE_DELETE_ON_CLOSE` / `DeleteFile` ?

**Reply by zqlustc** [reply]
yes, CreateFile -> SetRenameInformationFile

**Reply by Liryna** [reply]
SetRenameInformationFile for a delete ? That's surprising. Could you share the logs ?

**Reply by zqlustc** [reply]
I am sorry that I count not share the logs. In the windows NTFS, it's the `CreateFile -> SetRenameInformationFile` when i delete a file that I watched in Procmon

**Reply by Liryna** [reply]
I checked on my side and it just doing a CreateFile with the delete in close option. There is no rename to delete the file.

Also the code that you added in the driver can easily be handled in userland by denying the CreateFile with DeleteOnClose set. Like returning a status acces denied.

**Reply by kyanha** [reply]
> 
> 
> yes, CreateFile -> SetRenameInformationFile

If you get CreateFile -> SetRenameInformationFile, you're likely working from Windows Explorer, and you're almost definitely clicking the filename and pressing delete.  The fact that it's reappearing is likely because you can't rename a file to an empty name.

To select what you should be selecting, click the icon for the file, not the name of the file, and then press delete.

(Ideally, you'd use the command prompt and "delete filename.ext", but explaining how to do that is a bit on the complex side.)

In any case, the expected behavior to see in ProcMon is:
IRP_MJ_CREATE with "Desired Access: Delete, Disposition: Open, Options: Non-Directory File, Delete On Close, Attributes: n/a, ShareMode: Delete, AllocationSize: n/a, OpenResult: Opened"
IRP_MJ_CLEANUP
IRP_MJ_CLOSE

If you don't see that, then you aren't correctly invoking the file delete capability.

**Comment by kyanha** [comment, marked as answer]
By the way...

> ```
>     disposition = (irpSp->Parameters.Create.Options >> 24) & 0x000000ff;
>     if (irpSp->Parameters.Create.Options & FILE_DELETE_ON_CLOSE) {
>       status = STATUS_INVALID_PARAMETER;
>       __leave;
>     }
> ```

If you do this, then you will never be able to delete any file via the dokan driver that you compiled, because the file handle with the necessary permission and operation will never be opened, so it will never be closed and acted on.

**Reply by zqlustc** [reply]
Yes, thanks to your remiders. I have gaven up the minds.
